avocado/core/virt.py

```diff
@@ -268,34 +277,54 @@ def setup_login(self, hostname, username, password=None):
         else:
             self.logged = False
 
-    def ip_address(self):
+    def ip_address(self, timeout=30):
         """
-        Returns the IP address associated with the MAC address of this VM
-
-        Inspired from:
-        https://raw.githubusercontent.com/mcepl/virt-addr/master/virt-addr.py
+        Returns the IP address consulting libvirt domifaddr
 
         :returns: either the IP address or None if not found
         :rtype: str or None
         """
-        desc = etree.fromstring(self.domain.XMLDesc(0))
-        mac_path = "devices/interface[@type='network']/mac"
-        node = desc.find(mac_path)
-        if node is None:
-            return None
-
-        mac = node.get("address")
-        if mac is None:
-            return None
-
-        mac = mac.lower().strip()
-        output = subprocess.Popen(["arp", "-n"],
-                                  stdout=subprocess.PIPE).communicate()[0]
-        lines = [line.split() for line in output.split("\n")[1:]]
-        addresses = [line[0] for line in lines if (line and (line[2] == mac))]
-        if addresses:
-            # Just return the first address, this is a best effort attempt
-            return addresses[0]
+        timelimit = time.time() + timeout
+
+        # VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE: Parse DHCP lease file
+        # VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_AGENT: Ask Qemu Guest Agent
+        querytype = libvirt.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_AGENT
```

Comment:
With these changes, IMHO it doesn't make sense to have the two query types listed here.  You can even go one step further, and eliminate the `querytype` parameter.  I'd also name `_get_ip()` something like `_get_ip_from_libvirt_agent()`.